### PR TITLE
Implement global_asm!() (RFC 1548)

### PR DESCRIPTION
**Pull request by mrhota**
This is a first attempt. ~~One (potential) problem I haven't solved is how to handle multiple usages of `global_asm!` in a module/crate. It looks like `LLVMSetModuleInlineAsm` overwrites module asm, and `LLVMAppendModuleInlineAsm` is not provided in LLVM C headers 😦~~

I can provide more detail as needed, but honestly, there's not a lot going on here.

r? @eddyb 

CC @Amanieu @jackpot51 

Tracking issue: #35119

### CODE REVIEW

**Comment by mrhota** [comment]
I will rebase this evening.

**Comment by eddyb** [comment]
cc @nagisa @michaelwoerister

**Comment by mattico** [comment]
If an LLVM API is not provided by the C API, you can add a C++ wrapper in [rustllvm](https://github.com/rust-lang/rust/tree/master/src/rustllvm) and the corresponding Rust types in [librustc_llvm](https://github.com/rust-lang/rust/tree/master/src/librustc_llvm). Look at the rest of the file to see the conventions that are used (to_rust, from_rust, etc.)

**Comment by mrhota** [comment]
Thank you @mattico!

@steveklabnik I updated the unstable-book, so you might want to look it over.

File: src/doc/unstable-book/src/global_asm.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+A simple usage looks like this:
+
+```rust,ignore
+global_asm!(r#"
+    .global my_asm_func
+    .hidden __cancel
+my_asm_func:
+    ret
+__other_cancel:
+    jmp __cancel
```
Comment (parched) [inline src/doc/unstable-book/src/global_asm.md]: This would need to be mangled, wouldn't it?

Comment (mrhota) [inline src/doc/unstable-book/src/global_asm.md]: ah, yes, we need a `#[no_mangle]`. Thanks!

**Comment by mrhota** [comment]
We have a feature-gate test and a couple codegen tests! We also have a new unstable-book page. My initial problems/questions have been resolved.

File: src/doc/unstable-book/src/global_asm.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+symbols defined in the inline assembly, modulo any symbols marked as
+external. It also means syntax for directives and assembly follow the
+conventions of the assembler in your toolchain.
+
+A simple usage looks like this:
+
+```rust,ignore
+# #![feature(global_asm)]
+# you also need relevant target_arch cfgs
+global_asm!(include_str("something_neato.s"));
```
Comment (Amanieu) [inline src/doc/unstable-book/src/global_asm.md]: This should be `include_str!` (missing !).

**Review by Amanieu — CHANGES_REQUESTED** [review]
I think a test should be added which uses `include_str!` with `global_asm!`.

File: src/doc/unstable-book/src/global_asm.md (hunk 2 above)
Comment (mrhota) [inline src/doc/unstable-book/src/global_asm.md]: thanks

**Comment by mrhota** [comment]
@Amanieu we have a new test using `include_str!`. Hopefully it's what you had in mind.

**Comment by Amanieu** [comment]
@mrhota Yes, that's what I had in mind.

**Comment by mrhota** [comment]
@eddyb @nagisa @michaelwoerister any thoughts here? not sure who should review

File: src/test/codegen/global_asm.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-arm
+// ignore-aarch64
```
Comment (nagisa) [inline src/test/codegen/global_asm.rs]: Since assembly is for x86, this list should be significantly longer.

Namely we also support: MIPS, PowerPC, SystemZ, wasm, asmjs, nacl, nvptx, avr, msp430 and probably more stuff I forgot to mention.

File: src/test/codegen/global_asm_include.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-arm
+// ignore-aarch64
```
Comment (nagisa) [inline src/test/codegen/global_asm_include.rs]: Since assembly is for x86, this list should be significantly longer.

File: src/libsyntax/ast.rs [hunk 5]
```diff
@@ -1613,6 +1613,16 @@ pub struct ForeignMod {
     pub items: Vec<ForeignItem>,
 }
 
+/// Global inline assembly
+///
+/// aka module-level assembly or file-scoped assembly
+#[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug, Copy)]
+pub struct GlobalAsm {
+    pub asm: Symbol,
+    pub asm_str_style: StrStyle,
```
Comment (nagisa) [inline src/libsyntax/ast.rs]: Seems to be unused/unnecessary?

**Review by nagisa — COMMENTED** [review]
Seems good to me overall.

I’d like, however, more tests. Namely to ensure stuff like these *compile*. Compilation here is the important distinction, as codegen tests do not actually go through the assembler, and only generate LLVM-IR, which does not insure the actual codegen works.

```
// cfg-gating works properly
#[cfg(target_arch="x86")] global_asm! { ... }
#[cfg(target_arch="x86_64")] global_asm! { ... }
```
---

```
// codegen tests for all borderline T-1 architectures
global_asm!("foo blt foo") // arm
global_asm!("foo: j foo") // mips
global_asm!("foo: b foo") // ppc
```

File: src/test/codegen/global_asm.rs (hunk 3 above)
Comment (nagisa) [inline src/test/codegen/global_asm.rs]: A good approach would be to `git grep // ignore- src/tests` and add every single suffix that's seems unrelated to x86.

**Comment by mrhota** [comment]
@nagisa I might be missing something, but what's the point of testing full-compile _in addition_ to codegen? The real meat of this feature enables insertion of `module asm` into llvm-ir by using `global_asm!`, which the codegen tests ensure.

If we test full-compile, then we're just testing that LLVM does what it says it'll do with `module asm`s. If the compile fails at _that_ point, `rustc` is off the hook, since it's either an LLVM problem, or the programmer provided bad assembly.

```rust
global_asm!("put on your time travel trousers!");
```
should in fact go all the way _through_ codegen, and should produce

```
... module asm "put on your time travel trousers!"
```
which will obviously fail to assemble, but that's not our problem. The assembly template provided to `global_asm!` is (semantically) completely inaccessible to `rustc` and LLVM.

But, again, the codegen tests ensure `rustc` faithfully has handed the provided assembly template to LLVM, which itself promises faithfully to deliver it to the assembler.

File: src/libsyntax/ast.rs (hunk 5 above)
Comment (mrhota) [inline src/libsyntax/ast.rs]: good catch

File: src/test/codegen/global_asm_include.rs (hunk 4 above)
Comment (mrhota) [inline src/test/codegen/global_asm_include.rs]: I copied this from another codegen test, I think. I'll fill out the list.

**Comment by nagisa** [comment]
> If the compile fails at that point, rustc is off the hook, since it's either an LLVM problem, or the programmer provided bad assembly.

This is irrelevant from the perspective of the end user – they use rustc, not LLVM. You can assign blame to some dependency, but the fact of dependencies not working propagates to the end products.

Now, I don’t feel strongly about this, since this is will be unstable for a foreseeable future, but such tests will have to materialise before this gets stabilised. Might as well materialise them right now, then, right?

**Comment by mrhota** [comment]
@nagisa I'm at a loss; I have no idea why the builder keeps failing codegen tests since I added more `// ignore-`s. It passes locally.

**Comment by nagisa** [comment]
> I'm at a loss; I have no idea why the builder keeps failing codegen tests since I added more // ignore-s. It passes locally.

I’ve got no idea why it fails either. I’ll make a mental note to try and build this locally, but am likely to forget.

**Comment by mrhota** [comment]
@nagisa thanks. I ran the failing docker container's build on my mac and it passed. I ran the tests on my mac, too, and they pass.

**Comment by mrhota** [comment]
@nagisa OK, resolved build failure. Was that the last thing on your checklist?

@Amanieu github indicates that you requested changes. I think I resolved the issues you mentioned, but I don't know if you manually have to check the box...

**Comment by Amanieu** [comment]
I don't see a box to check. Anyways, I'm happy with the changes.

**Comment by mrhota** [comment]
@nagisa what are we waiting on?

**Comment by mrhota** [comment]
@arielb1 `yum` failures on builder; not compilation failure

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/issues/40474

**Comment by mrhota** [comment]
@alexcrichton failure due to some ssl error:

```
Build started
git clone -q --depth=1 --branch=auto https://github.com/rust-lang/rust.git C:\projects\rust
fatal: unable to access 'https://github.com/rust-lang/rust.git/': Unknown SSL protocol error in connection to github.com:443 
Command exited with code 128
cat %CD%/sccache.log
cat: 'C:\projects\rust/sccache.log': No such file or directory
Command exited with code 1
```

**Comment by alexcrichton** [comment]
@bors: retry

* network failure https://github.com/rust-lang/rust/issues/40474

**Comment by arielb1** [comment]
```
error[E0412]: cannot find type `ExpnId` in this scope
    --> src\libsyntax\ast.rs:1592:18
     |
1592 |     pub expn_id: ExpnId,
     |                  ^^^^^^ not found in this scope
error[E0412]: cannot find type `ExpnId` in this scope
    --> src\libsyntax\ast.rs:1592:18
     |
1592 |     pub expn_id: ExpnId,
     |                  ^^^^^^ not found in this scope
error[E0412]: cannot find type `ExpnId` in this scope
    --> src\libsyntax\ast.rs:1592:18
     |
1592 |     pub expn_id: ExpnId,
     |                  ^^^^^^ not found in this scope
   Compiling rls-span v0.1.0
error[E0204]: the trait `Copy` may not be implemented for this type
    --> src\libsyntax\ast.rs:1589:77
     |
1589 | #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug, Copy)]
     |                                                                             ^^^^
...
1592 |     pub expn_id: ExpnId,
     |     ------------------- this field does not implement `Copy`
error: aborting due to previous error
   Compiling arena v0.0.0 (file:///C:/projects/rust/src/libarena)
error: Could not compile `syntax`.
Build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by mrhota** [comment]
@arielb1 failure fixed

**Comment by frewsxcv** [comment]
@bors r-

This hit a failure on windows:

https://github.com/rust-lang/rust/pull/41115

**Comment by mrhota** [comment]
@frewsxcv Not sure what happened on windows. I can `#[allow(dead_code)]` on the `foo` declaration to get rid of the `#[warn(dead_code)]`, but the linker croaked: could windows linker be wiping that symbol out since nothing calls it? Why doesn't that happen on other platforms?

**Comment by mrhota** [comment]
@michaelwoerister can you review my changes wrt your ICH work?

@frewsxcv other than that, I think this is ready to go, again. I can't test on Windows, so if someone with experience on that platform can give me advice, that would be appreciated.

**Comment by michaelwoerister** [comment]
Looks great to me, both with respect to ICH computation and translation item collection. Excellent work!

If you wanted to be super-thorough you could add `codegen-units` tests that make sure that the translation item collector finds the item and puts it into the right codegen unit.

**Comment by mrhota** [comment]
@frewsxcv @michaelwoerister this is ready, I think. Are we waiting on something else?

@michaelwoerister I will add a test in the spirit of your suggestion in a later PR if that's alright

**Comment by aturon** [comment]
@bors r=mw

(Taking the liberty of allowing the extra test to land separately)

**Comment by michaelwoerister** [comment]
Just to be clear, I only reviewed the ICH and TransItem parts; @nagisa reviewed the rest. I'm fine with the codegen units tests being added later.

File: src/test/run-pass/simple_global_asm.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(global_asm)]
+#![feature(naked_functions)]
+
+#[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
```
Comment (TimNN) [inline src/test/run-pass/simple_global_asm.rs]: This fails on eg. ARM, see https://github.com/rust-lang/rust/pull/41242#issuecomment-293508331

**Comment by TimNN** [comment]
@bors r-

(See inline comment)

**Comment by mrhota** [comment]
@TimNN ready to rock again

**Comment by TimNN** [comment]
@bors retry

`fatal: unable to access 'https://github.com/rust-lang/rust.git/': Unknown SSL protocol error in connection to github.com:443` ?!

**Comment by mrhota** [comment]
same SSL error on Windows

**Comment by alexcrichton** [comment]
@bors: retry

* appveyor git failure - https://github.com/rust-lang/rust/issues/40474

**Comment by frewsxcv** [comment]
That appveyor failure looks legit.

**Comment by mrhota** [comment]
```
failures:
---- [run-pass] run-pass\simple_global_asm.rs stdout ----
	
error: compilation failed!
status: exit code: 101
command: PATH="C:\projects\rust\build\i686-pc-windows-gnu\stage2\bin;C:\projects\rust\build\i686-pc-windows-gnu\stage2-tools\i686-pc-windows-gnu\release\deps;C:\projects\rust\build\i686-pc-windows-gnu\stage2\lib\rustlib\i686-pc-windows-gnu\lib;C:\Program Files (x86)\Inno Setup 5;C:\Python27;C:\projects\rust\mingw32\bin;C:\msys64\usr\bin;C:\Perl\site\bin;C:\Perl\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Program Files\7-Zip;C:\Program Files\Microsoft\Web Platform Installer;C:\Tools\GitVersion;C:\Tools\PsTools;C:\Program Files\Git LFS;C:\Program Files (x86)\Subversion\bin;C:\Program Files\Microsoft SQL Server\120\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\110\Tools\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\Tools\Binn;C:\Program Files\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\Tools\Binn\ManagementStudio;C:\Tools\WebDriver;C:\Program Files (x86)\Microsoft SDKs\TypeScript\1.4;C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\PrivateAssemblies;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\wbin;C:\Ruby193\bin;C:\Tools\NUnit\bin;C:\Tools\xUnit;C:\Tools\MSpec;C:\Tools\Coverity\bin;C:\Program Files (x86)\CMake\bin;C:\go\bin;C:\Program Files\Java\jdk1.8.0\bin;C:\Python27;C:\Program Files\nodejs;C:\Program Files (x86)\iojs;C:\Program Files\iojs;C:\Users\appveyor\AppData\Roaming\npm;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files (x86)\MSBuild\14.0\Bin;C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\Microsoft\SQLDB\DAC\120;C:\Tools\NuGet;C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\Program Files\Microsoft DNX\Dnvm;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\130\Tools\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Apache\Maven\bin;C:\Python27\Scripts;C:\Tools\NUnit3;C:\ProgramData\chocolatey\bin;C:\Program Files\Mercurial;C:\Program Files (x86)\Yarn\bin;C:\Program Files\LLVM\bin;C:\Program Files\dotnet;C:\Program Files\erl8.3\bin;C:\Tools\curl\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files (x86)\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\usr\bin;C:\Program Files\Amazon\AWSCLI;C:\Users\appveyor\AppData\Local\Yarn\.bin;C:\Users\appveyor\AppData\Roaming\npm;C:\Program Files\AppVeyor\BuildAgent;C:\projects\rust;C:\projects\rust\handle" C:\projects\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe C:\projects\rust\src/test\run-pass\simple_global_asm.rs -L C:\projects\rust\build\i686-pc-windows-gnu\test\run-pass --target=i686-pc-windows-gnu --error-format json -L C:\projects\rust\build\i686-pc-windows-gnu\test\run-pass\simple_global_asm.stage2-i686-pc-windows-gnu.run-pass.libaux -C prefer-dynamic -o C:\projects\rust\build\i686-pc-windows-gnu\test\run-pass\simple_global_asm.stage2-i686-pc-windows-gnu.exe -Crpath -O -Lnative=C:\projects\rust\build\i686-pc-windows-gnu\native\rust-test-helpers
stdout:
------------------------------------------
------------------------------------------
stderr:
------------------------------------------
{"message":"linking with `gcc` failed: exit code: 1","code":null,"level":"error","spans":[],"children":[{"message":"\"gcc\" \"-Wl,--enable-long-section-names\" \"-fno-use-linker-plugin\" \"-Wl,--nxcompat\" \"-nostdlib\" \"-Wl,--large-address-aware\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\\\\crt2.o\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\\\\rsbegin.o\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\test\\\\run-pass\\\\simple_global_asm.0.o\" \"-o\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\test\\\\run-pass\\\\simple_global_asm.stage2-i686-pc-windows-gnu.exe\" \"-Wl,--gc-sections\" \"-nodefaultlibs\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\test\\\\run-pass\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\test\\\\run-pass\\\\simple_global_asm.stage2-i686-pc-windows-gnu.run-pass.libaux\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\native\\\\rust-test-helpers\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\" \"-L\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\" \"-l\" \"std-4d6881ec6132b951\" \"-Wl,-Bstatic\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\\\\libcompiler_builtins-7ac5a34e9b48514f.rlib\" \"-Wl,-Bdynamic\" \"-l\" \"advapi32\" \"-l\" \"ws2_32\" \"-l\" \"userenv\" \"-l\" \"shell32\" \"-lmingwex\" \"-lmingw32\" \"-lgcc\" \"-lmsvcrt\" \"-luser32\" \"-lkernel32\" \"C:\\\\projects\\\\rust\\\\build\\\\i686-pc-windows-gnu\\\\stage2\\\\lib\\\\rustlib\\\\i686-pc-windows-gnu\\\\lib\\\\rsend.o\"","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"C:\\projects\\rust\\build\\i686-pc-windows-gnu\\test\\run-pass\\simple_global_asm.0.o:(.text+0x15): undefined reference to `foo'\r\ncollect2.exe: error: ld returned 1 exit status\r\n","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":null}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":null}
------------------------------------------
thread '[run-pass] run-pass\simple_global_asm.rs' panicked at 'explicit panic', src\tools\compiletest\src\runtest.rs:2627
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    [run-pass] run-pass\simple_global_asm.rs
test result: FAILED. 2632 passed; 1 failed; 17 ignored; 0 measured
```

@alexcrichton @nagisa any pointers? I don't have access to a `MSYS_BITS=32, RUST_CONFIGURE_ARGS=--build=i686-pc-windows-gnu` environment. Maybe there's something obvious I missed, but I don't see how `foo` should be undefined in `src/test/run-pass/simple_global_asm.rs` on that environment.

Also, note that `codegen` tests pass on that platform, which is to say `global_asm`'s argument is faithfully passed to LLVM and written out in llvm-ir, which is the whole point of this feature. A test like this was my worry when, as I indicated to @nagisa before, this test is now _only_ testing that platform's assembler and linker, not `rustc`. I'm now caught in an endless loop of fiddling with a test until it passes. That is, it's a pointless test.

**Comment by alexcrichton** [comment]
The failure here is:

> undefined reference to 'foo'

That's probably because of the way symbols are named across systems. I believe on some platforms (like OSX and Windows) you need a leading underscore (e.g. `_foo` in the asm) whereas others like Linux you can omit it.

**Comment by mrhota** [comment]
@alexcrichton that seems to be it. Thanks!

**Comment by mrhota** [comment]
@alexcrichton ready again...

I declared both `foo` and `_foo` in the assembly. Hopefully this will satisfy windows' linker:
```
    .global foo
    .global _foo
foo:
_foo:
    ret
```
